File: coolrmi/hu.qgears.coolrmi/src/hu/qgears/coolrmi/serializer/TypeSerializer.java
package hu.qgears.coolrmi.serializer;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public abstract class TypeSerializer {
	final TypeId type;
	final Class<?> serializedClass, primitiveClass;
	private final JavaType javaType;

	public TypeSerializer(TypeId type, Class<?> serializedClass,
			Class<?> primitveClass) {
		this.type = type;
		this.serializedClass = serializedClass;
		this.primitiveClass = primitveClass;
		this.javaType = new JavaType(
				primitiveClass != null ? primitiveClass : serializedClass);
	}

	public void writeType(PortableSerializer serializer, OutputStream os,
			JavaType typ) throws IOException {
		os.write(type.ordinal());
	}

	public JavaType readType(PortableSerializer serializer, InputStream is)
			throws IOException, ClassNotFoundException {
		return javaType;
	}

	public abstract void serialize(PortableSerializer serializer,
			OutputStream os, Object o, JavaType typ) throws IOException;

	public abstract Object deserialize(PortableSerializer serializer,
			InputStream is, JavaType typ) throws Exception;

	public boolean canSerializeIsSpecial() {
		return false;
	}

	public boolean canSerialize(PortableSerializer serializer, JavaType typ) {
		return typ.getCls().equals(serializedClass)
				|| typ.getCls().equals(primitiveClass);
	}

	public boolean isPolymorphic() {
		return false;
	}
}

File: coolrmi/CoolRMI.Net/Serializer/TypeSerializer.cs
using System;
using System.IO;

namespace CoolRMI.Net.Serializer
{
    public enum TypeId
    {
        Null,
        Bool,
        Int8,
        Int16,
        Int32,
        Int64,
        Char,
        Float,
        Double,
        Enum,
        String,
        Object,
        Array,
        Exception,
        List,
    }

    public abstract class TypeSerializer
    {
        public TypeId Type { get; }
        public Type DotNetType { get; }

        protected TypeSerializer(TypeId type, Type dotNetType)
        {
            Type = type;
            DotNetType = dotNetType;
        }

        public virtual void WriteType(PortableSerializer serializer,
            BinaryWriter bw, Type typ)
        {
            bw.Write((byte) Type);
        }

        public virtual Type ReadType(PortableSerializer serializer,
            BinaryReader br)
        {
            return DotNetType;
        }

        public abstract void Serialize(PortableSerializer serializer,
            BinaryWriter bw, object o, Type typ);

        public abstract object Deserialize(PortableSerializer serializer,
            BinaryReader br, Type typ);

        public virtual bool CanSerializeIsSpecial => false;

        public virtual bool CanSerialize(PortableSerializer serializer, Type typ)
        {
            return typ == DotNetType;
        }

        public virtual bool IsPolymorphic => false;
    }
}
